proxy/client.go

```diff
@@ -0,0 +1,233 @@
+package proxy
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+)
+
+const (
+	// timeout for upstream /sync requests (after which it will send back
+	// an empty response)
+	syncTimeout = 60 * time.Second
+)
+
+type MatrixClient struct {
+	AccessToken string
+
+	url string
+
+	NextSyncBatch string
+
+	// our client for the upstream connection
+	httpClient http.Client
+}
+
+func NewClient(url string, accessToken string) *MatrixClient {
+	if !strings.HasSuffix(url, "/") {
+		url += "/"
+	}
+	return &MatrixClient{url: url, AccessToken: accessToken}
+}
+
+// an error returned when a matrix endpoint returns a non-200 whose body is
+// not a matrix error
+type HttpError struct {
+	StatusCode  int
+	ContentType string
+	Body        []byte
+}
+
+func (s *HttpError) Error() string {
+	return string(s.Body)
+}
+
+// an error returned when a matrix endpoint returns a standard matrix error
+type MatrixError struct {
+	HttpError
+	Details MatrixErrorDetails
+}
+
+func (e *MatrixError) Error() string {
+	return fmt.Sprintf("%s (%s)", e.Details.ErrCode, e.Details.Error)
+}
+
+type MatrixErrorDetails struct {
+	ErrCode string `json:"errcode"`
+	Error   string `json:"error"`
+}
+
+// Sync sends the sync request, and returns the body of the response,
+// or an error.
+//
+// It keeps track of the 'next_batch' from the result, and uses it to se the
```

Comment:
Worth documenting the behaviour for next_batch if an error is encountered (i.e. that it will retry the previous request, rather than skipping them).